feature.py

```diff
@@ -67,57 +67,75 @@ def forward(self, x: th.Tensor) -> th.Tensor:
 class Residual(nn.Module):
     """CBA + Residual Block."""
 
-    def __init__(self, base: nn.Module):
-        self.base = base
+    def __init__(
+        self, 
+        c_in: int, 
+        c_out: int, 
+        padding: Union[int, Tuple[int, int]] = 1, 
+        stride: Union[int, Tuple[int, int]] = 1
+    ):
+        super().__init__()
+        self.conv1 = CBA(c_in, c_out, padding=padding, stride=stride)
+        self.conv2 = CBA(c_out, c_out, lrelu_eps=None)
+        self.residual = nn.Sequential()
+        if c_in != c_out or stride != 1:
+            self.residual = CBA(c_in, c_out, padding=padding, stride=stride, lrelu_eps=None)
 
     def forward(self, x: th.Tensor) -> th.Tensor:
-        return self.base(x) + self.residual(x)
+        out = self.conv1(x)
+        out = self.conv2(out)
+        out += self.residual(x)
+        out = F.leaky_relu(out)
+        return out
 
 
 class GlyphEncoder(nn.Module):
     """Glyph (basically, the map/env) encoder."""
 
-    def __init__(self):
+    def __init__(self, in_channels: List[int], out_channels: List[int]):
         super().__init__()
 
-        # Most naive implementation in the world
         self.f = nn.Sequential(*[
-            CBA(1, 16, (3, 3), stride=(3, 3), padding=(0, 1)),
-            CBA(16, 32, (3, 3), padding=0),
-            CBA(32, 64, (3, 3), padding=0),
-            nn.Flatten()
+            # -- [K x 79 x 21]
+            Residual(in_channels[0], out_channels[0]),
+            # -- [A x 27 x 7]
+            Residual(in_channels[1], out_channels[1], padding=(1, 0), stride=3),
+            # -- [A x 9 x 3]
+            CBA(in_channels[2], out_channels[2], padding=(0, 1), stride=3),
+            # -- [A x 27]
+            nn.Flatten(start_dim=2)
         ])
 
     def forward(self, x: th.Tensor) -> th.Tensor:
         """
         Args:
-            x: int16 tensor of shape (..., 21, 79),
-               in range (0, 5976).
+            x: tensor of shape (..., embedding_dim, 21, 79).
         Returns:
-            y: (..., D) feature embedding of `x`.
+            y: (..., A, 27) feature embedding of `x`.
```

Comment:
What is `A`? Is it just some arbitrary dimension? Or does it have some meaning?

Reply:
A stands for Attention. Dimension 'A' is used for query.